### PR TITLE
cctools: move to github

### PR DESCRIPTION
**Pull request by michaelld**
#### Description

cctools: move to github

also fixup patches to apply cleanly

Closes: https://trac.macports.org/ticket/66436

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 11.7.3 20G1102 x86_64
Xcode 12.3 12C33

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Review by kencu — APPROVED** [review]
lgtm!

File: devel/cctools/files/cctools-829-lto.patch [hunk 1]
```diff
@@ -14,7 +14,7 @@
 -				    "xctoolchain/usr/lib/libLTO.dylib",
 -				    RTLD_NOW);
 +		const char *lto_paths[] = {
-+		    "@@LLVM_LIBDIR@@/libLTO.dylib",
++		    "/opt/local/libexec/llvm-11/lib/libLTO.dylib",
```
Comment (Gcenx) [inline devel/cctools/files/cctools-829-lto.patch:17]: Assuming you didn’t mean to hardcore the llvm version here

Comment (kencu) [reply devel/cctools/files/cctools-829-lto.patch:17]: oh, good pickup

Comment (michaelld) [reply devel/cctools/files/cctools-829-lto.patch:17]: Good catch! No that's not intentional... guessing it's patched by a `reinplace` & I somehow missed that! I won't be able to fix this until Friday as I'm traveling without a work computer. If someone else wants to fix it go do it!

**Comment by michaelld** [comment]
hmmm ... updated the branch & GitHub decided to close this ... I don't think this PR was merged ... will check again & see if I can revive this PR & if not (both) then I'll do another PR

**Comment by michaelld** [comment]
OK worked out the issue & reopened this with the fixed patchfile per the prior comment.

**Review by Gcenx — APPROVED** [review]
LGTM

**Comment by kencu** [comment]
I'm trying to fix tcp_wrappers, which has the same issue...

do we need to do the stealth-update dance here? The files have the same name name as before, but yet are different, right?

**Comment by michaelld** [comment]
Yes ... this is basically a stealth update ... good catch! Will update this PR shortly

**Comment by michaelld** [comment]
Added stealth update code ... let's see how checks do

**Comment by michaelld** [comment]
All checks look good & I've built this port on multiple different OSX now ... so _I_ think it's good to go

**Review by Gcenx — APPROVED** [review]
Looks good to me.

**Comment by catap** [comment]
I can confirm that it works for me on 10.5

**Comment by catap** [comment]
@michaelld can you try to fetch sources right now? `https://github.com/apple-oss-distributions/cctools/archive/949.0.1/cctools-949.0.1.tar.gz` leads to `404` right now from my place.

:(

**Comment by Gcenx** [comment]
The url should be `https://github.com/apple-oss-distributions/cctools/archive/refs/tags/cctools-949.0.1.tar.gz`

File: devel/cctools/Portfile [hunk 2]
```diff
@@ -20,19 +24,23 @@ long_description        A set of essential tools to support development \
...
-checksums               ld64-530.tar.gz \
-                        rmd160  793d012a611de5a044db3f97c6a9ec149072aff6 \
-                        sha256  ee37f0487601c08c7d133bc91cad2e9084d00d02aa4709d228a9a065960aa187 \
-                        size    744871 \
-                        cctools-949.0.1.tar.gz \
+distfiles               ${name}-${version}.tar.gz:main \
+                        ld64-${ld64_version}.tar.gz:ld64
+
+master_sites            ${github.homepage}/archive/${git.branch}:main \
+                        https://github.com/${github.author}/ld64/archive/ld64-${ld64_version}:ld64
```
Comment (catap) [inline devel/cctools/Portfile:34]: @Gcenx but it is hardcoded on the different way here.

File: devel/cctools/Portfile [hunk 3]
```diff
@@ -20,19 +24,23 @@ long_description        A set of essential tools to support development \
                         to binutils on other platforms.
 
 homepage                https://opensource.apple.com/source/${name}/
-master_sites            https://opensource.apple.com/tarballs/${name}:main \
-                        https://opensource.apple.com/tarballs/ld64:ld64
 
-distfiles               ld64-${ld64_version}.tar.gz:ld64 ${name}-${version}.tar.gz:main
+github.setup            apple-oss-distributions ${name} ${version}
```
Comment (catap) [inline devel/cctools/Portfile:28]: I'm wonder why have you used github PG if you're adding homepage and master_sites by userself?

**Comment by catap** [comment]
BTW if I'm witching to `https://github.com/apple-oss-distributions/cctools/archive/refs/tags/cctools-949.0.1.tar.gz` it fails to build as: `Error: Failed to extract cctools: error copying "/opt/local/var/macports/build/...work/cctools-949.0.1/../ld64-530/src/other/PruneTrie.cpp": no such file or directory`

:(

**Comment by kencu** [comment]
yes, I"m getting a 404 for this too:

https://github.com/apple-oss-distributions/cctools/archive/949.0.1/cctools-949.0.1.tar.gz

and when I change it to this:

master_sites            https://github.com/apple-oss-distributions/cctools/archive/refs/tags:main \

I get a checksum error, and when I fix that, I get the same error as @catap mentioned:

```
--->  Extracting cctools
--->  Extracting cctools-949.0.1.tar.gz
Executing:  cd "/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_devel_cctools/cctools/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/cctools/949.0.1_1/cctools-949.0.1.tar.gz' | /usr/bin/tar -xf - 
--->  Extracting ld64-530.tar.gz
Executing:  cd "/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_devel_cctools/cctools/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/cctools/949.0.1_1/ld64-530.tar.gz' | /usr/bin/tar -xf - 
Error: Failed to extract cctools: error copying "/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_devel_cctools/cctools/work/cctools-949.0.1/../ld64-530/src/other/PruneTrie.cpp": no such file or directory
Error: See /opt/local/var/macports/logs/_opt_macportsofficial_macports-ports_devel_cctools/cctools/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there is a bug.
Error: Processing of port cctools failed
```

which looks to me like it has something to do with some weird archive names?

```
$ ls
cctools-cctools-949.0.1	ld64-ld64-530
```

this new Apple OSS on github stuff is going to need some straightening out. Maybe it's own entry in the GitHub PG? I dunno.

**Comment by kencu** [comment]
so fixing this weird archive folder names with a (who-knows-why-needed?) symlink, it works. This was the grand total hacky patch I used to get it to build (yuk):

```
$ git diff
diff --git a/devel/cctools/Portfile b/devel/cctools/Portfile
index b5d70b09748..8c882db2b37 100644
--- a/devel/cctools/Portfile
+++ b/devel/cctools/Portfile
@@ -30,17 +30,17 @@ github.setup            apple-oss-distributions ${name} ${version}
 distfiles               ${name}-${version}.tar.gz:main \
                         ld64-${ld64_version}.tar.gz:ld64
 
-master_sites            ${github.homepage}/archive/${git.branch}:main \
-                        https://github.com/${github.author}/ld64/archive/ld64-${ld64_version}:ld64
+master_sites            https://github.com/apple-oss-distributions/cctools/archive/refs/tags:main \
+                        https://github.com/apple-oss-distributions/ld64/archive/refs/tags:ld64
 
 checksums               cctools-949.0.1.tar.gz \
-                        rmd160  6412d9c5aea6aadd88cd9abdde6c1383f103b582 \
-                        sha256  830485ac7c563cd55331f643952caab2f0690dfbd01e92eb432c45098b28a5d0 \
-                        size    1968671 \
+                        rmd160  54a8e318087547b827eee4e5dbb21a8f3956402f \
+                        sha256  8b2d8dc371a57e42852fa6102efaf324ef004adf86072bf9957e2ac9005326c1 \
+                        size    1967695 \
                         ld64-530.tar.gz \
-                        rmd160  793d012a611de5a044db3f97c6a9ec149072aff6 \
-                        sha256  ee37f0487601c08c7d133bc91cad2e9084d00d02aa4709d228a9a065960aa187 \
-                        size    744871
+                        rmd160  57d6629f31180b53f6f6a324a28e723611bd3cfb \
+                        sha256  a21de4c2d4cd3ebf13c887eebbafc4b58ff0ee5bb07e592ba12690fb28a9cb17 \
+                        size    729897
 
 patchfiles              cctools-829-lto.patch \
                         PR-37520.patch \
@@ -201,6 +201,8 @@ if {![variant_isset xcode] && ![variant_isset xtools]} {
 
 post-extract {
     if {![variant_isset xcode] && ![variant_isset xtools]} {
+        ln -s ${workpath}/ld64-ld64-${ld64_version} ${workpath}/ld64-${ld64_version}
+        ln -s ${workpath}/cctools-cctools-${version} ${workpath}/cctools-${version}
         file copy ${worksrcpath}/../ld64-${ld64_version}/src/other/PruneTrie.cpp ${worksrcpath}/misc
         system "touch ${worksrcpath}/../ld64-${ld64_version}/src/abstraction/configure.h"
     }
```

**Comment by kencu** [comment]
well, the weird name comes from upstream; when you download the file even with a web browser, you get this:

cctools-cctools-949.0.1.tar.gz

**Comment by catap** [comment]
@michaelld maybe you use Ken's patch and this PR will be merged? :)

**Comment by kencu** [comment]
this link does work, and it's more "MacPortsy":

https://github.com/apple-oss-distributions/cctools/archive/cctools-949.0.1.tar.gz

**Comment by kencu** [comment]
updated patch:
```
 % git diff
diff --git a/devel/cctools/Portfile b/devel/cctools/Portfile
index b5d70b09748..d5c1b683641 100644
--- a/devel/cctools/Portfile
+++ b/devel/cctools/Portfile
@@ -30,17 +30,17 @@ github.setup            apple-oss-distributions ${name} ${version}
 distfiles               ${name}-${version}.tar.gz:main \
                         ld64-${ld64_version}.tar.gz:ld64
 
-master_sites            ${github.homepage}/archive/${git.branch}:main \
+master_sites            ${github.homepage}/archive/:main \
                         https://github.com/${github.author}/ld64/archive/ld64-${ld64_version}:ld64
 
 checksums               cctools-949.0.1.tar.gz \
-                        rmd160  6412d9c5aea6aadd88cd9abdde6c1383f103b582 \
-                        sha256  830485ac7c563cd55331f643952caab2f0690dfbd01e92eb432c45098b28a5d0 \
-                        size    1968671 \
+                        rmd160  54a8e318087547b827eee4e5dbb21a8f3956402f \
+                        sha256  8b2d8dc371a57e42852fa6102efaf324ef004adf86072bf9957e2ac9005326c1 \
+                        size    1967695 \
                         ld64-530.tar.gz \
-                        rmd160  793d012a611de5a044db3f97c6a9ec149072aff6 \
-                        sha256  ee37f0487601c08c7d133bc91cad2e9084d00d02aa4709d228a9a065960aa187 \
-                        size    744871
+                        rmd160  57d6629f31180b53f6f6a324a28e723611bd3cfb \
+                        sha256  a21de4c2d4cd3ebf13c887eebbafc4b58ff0ee5bb07e592ba12690fb28a9cb17 \
+                        size    729897
 
 patchfiles              cctools-829-lto.patch \
                         PR-37520.patch \
@@ -56,6 +56,12 @@ patchfiles              cctools-829-lto.patch \
                         cctools-949-missed-stdio.diff \
                         cctools-949-lipo-segalign-log2fix.diff
 
+
+post-extract {
+    move ${workpath}/cctools-cctools-${version} ${workpath}/cctools-${version}
+    move ${workpath}/ld64-ld64-${ld64_version}  ${workpath}/ld64-${ld64_version}
+}
+
 # lipo-segalign-log2fix https://trac.macports.org/ticket/63164
 
 # small fixes and sdk adjustments made to 949 but
```

**Comment by michaelld** [comment]
I've integrated @kencu 's patch into this PR

File: devel/cctools/Portfile [hunk 4]
```diff
@@ -48,6 +56,18 @@ patchfiles              cctools-829-lto.patch \
...
+    move ${workpath}/cctools-cctools-${version} ${workpath}/cctools-${version}
+    move ${workpath}/ld64-ld64-${ld64_version}  ${workpath}/ld64-${ld64_version}
+}
+
+
+post-extract {
+    move ${workpath}/cctools-cctools-${version} ${workpath}/cctools-${version}
+    move ${workpath}/ld64-ld64-${ld64_version}  ${workpath}/ld64-${ld64_version}
+}
+
```
Comment (catap) [inline devel/cctools/Portfile:70]: @michaelld seems that the second block was added by mistake.

Comment (michaelld) [reply devel/cctools/Portfile:68]: Yup duplicate block ... fixed! Also with the current form of `extract.rename`, need to explicitly set it to `no` so I added that

**Comment by michaelld** [comment]
OK one more try ...

File: devel/cctools/Portfile (hunk 3 above)
Comment (michaelld) [reply devel/cctools/Portfile:28]: ... at least for the downloads .. the homepage is to override the GH PG default

Comment (catap) [reply devel/cctools/Portfile:28]: @michaelld yeah, and I think that if you remove that PG, you won't lose anything :)

**Comment by catap** [comment]
@michaelld tests had failed as:
```
Error: Failed to extract cctools: error renaming "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_devel_cctools/cctools/work/cctools-cctools-949.0.1": no such file or directory
```

**Comment by kencu** [comment]
something quite weird is going on here... the name of the downloaded file is changed on downloading it... (sometimes?) 

I'm not fully understanding what and why..

**Comment by edwardgalligan** [comment]
Found this PR from trac after encountering issues installing `cctools`.

I checked out this branch & ran `sudo port -vst install` which seemed to work fine - is it possible the build issues on Jan 16th were transient bugs on the tarball hosting end that have since been resolved?

**Comment by catap** [comment]
@michaelld do you plan to finish it? If no I may help.

**Comment by catap** [comment]
Here the fixed version of that PR: https://github.com/macports/macports-ports/pull/18443

I hope that @michaelld  doesn't mind.

**Comment by catap** [comment]
I guess this one should be closed.

**Pull request closed without merging**